p2p/pex/addrbook.go

```diff
@@ -324,10 +329,24 @@ func (a *addrBook) MarkAttempt(addr *p2p.NetAddress) {
 	ka.markAttempt()
 }
 
-// MarkBad implements AddrBook. Currently it just ejects the address.
-// TODO: black list for some amount of time
+// MarkBad implements AddrBook. Kicks address out from book, places
+// the address in the badPeers pool.
 func (a *addrBook) MarkBad(addr *p2p.NetAddress) {
-	a.RemoveAddress(addr)
+	if a.addBadPeer(addr) {
+		a.RemoveAddress(addr)
+	}
+}
+
+func (a *addrBook) ReinstateBadPeers() {
+	for _, ka := range a.badPeers {
+		if !ka.isBanned(defaultBanTime) {
+			a.mtx.Lock()
```

Comment:
The mutex lock must be taken out at the start of the function call, otherwise a concurrent thread can modify `badPeers` while you are iterating over it and then bad things happen.